git diff a4750b7a81c1ff2889d6d6edee5efe10343b2acf
diff --git a/target/mips/cheri-archspecific-early.h b/target/mips/cheri-archspecific-early.h
index c3941faa..232a4f5e 100644
--- a/target/mips/cheri-archspecific-early.h
+++ b/target/mips/cheri-archspecific-early.h
@@ -103,7 +103,7 @@ static inline const cap_register_t *cheri_get_ddc(CPUMIPSState *env) {
     return &env->active_tc.CHWR.DDC;
 }
 
-static inline const cap_register_t *_cheri_get_pcc_unchecked(CPUMIPSState *env)
+static inline const cap_register_t *_cheri_get_pcc_unchecked(const CPUMIPSState *env)
 {
     cheri_debug_assert(env->active_tc.PCC.cr_extra == CREG_FULLY_DECOMPRESSED);
     return &env->active_tc.PCC;
diff --git a/target/mips/cpu.h b/target/mips/cpu.h
index 1ff60dfb..247d7652 100644
--- a/target/mips/cpu.h
+++ b/target/mips/cpu.h
@@ -1504,7 +1504,7 @@ static inline target_ulong cpu_get_recent_pc(CPUMIPSState *env)
 #endif
 }
 
-static inline bool pc_is_current(CPUArchState *env)
+static inline bool pc_is_current(const CPUArchState *env)
 {
 #ifdef CONFIG_DEBUG_TCG
     return env->active_tc._pc_is_current;
